freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/migration/handler/DeleteCloudFormationHandler.java

```diff
@@ -0,0 +1,103 @@
+import com.sequenceiq.cloudbreak.cloud.handler.CloudPlatformEventHandler;
+import com.sequenceiq.cloudbreak.cloud.model.CloudCredential;
+import com.sequenceiq.cloudbreak.cloud.model.CloudResource;
+import com.sequenceiq.cloudbreak.cloud.model.CloudResourceStatus;
+import com.sequenceiq.cloudbreak.cloud.model.CloudStack;
+import com.sequenceiq.cloudbreak.cloud.notification.ResourceNotifier;
+import com.sequenceiq.cloudbreak.cloud.notification.model.ResourcePersisted;
+import com.sequenceiq.cloudbreak.cloud.service.ResourceRetriever;
+import com.sequenceiq.common.api.type.CommonStatus;
+import com.sequenceiq.common.api.type.ResourceType;
+
+import reactor.bus.Event;
+import reactor.bus.EventBus;
+
+@Component
+public class DeleteCloudFormationHandler implements CloudPlatformEventHandler<DeleteCloudFormationRequest> {
+
+    private static final Logger LOGGER = getLogger(DeleteCloudFormationHandler.class);
+
+    @Inject
+    private EventBus eventBus;
+
+    @Inject
+    private AwsTerminateService awsTerminateService;
+
+    @Inject
+    private AwsAuthenticator awsAuthenticator;
+
+    @Inject
+    private ResourceRetriever resourceRetriever;
+
+    @Inject
+    private ResourceNotifier resourceNotifier;
+
+    @Inject
+    private AwsMigrationUtil awsMigrationUtil;
+
+    @Override
+    public Class<DeleteCloudFormationRequest> type() {
+        return DeleteCloudFormationRequest.class;
+    }
+
+    @Override
+    public void accept(Event<DeleteCloudFormationRequest> event) {
+        LOGGER.info("Delete the old cloud formation during the AWS migration");
+        DeleteCloudFormationRequest request = event.getData();
+        try {
+            CloudContext cloudContext = request.getCloudContext();
+            CloudCredential cloudCredential = request.getCloudCredential();
+            CloudStack cloudStack = request.getCloudStack();
+            AuthenticatedContext ac = awsAuthenticator.authenticate(cloudContext, cloudCredential);
+            Optional<CloudResource> cfCloudResource = resourceRetriever
+                    .findFirstByStatusAndTypeAndStack(CommonStatus.CREATED, ResourceType.CLOUDFORMATION_STACK, request.getResourceId());
+            boolean cloudFormationTemplateDeleted = false;
+            if (cfCloudResource.isPresent()) {
+                if (awsMigrationUtil.allInstancesDeletedFromCloudFormation(ac, cfCloudResource.get())) {
+                    terminateCfStackAndResource(cloudContext, cloudStack, ac, cfCloudResource.get());
+                    cloudFormationTemplateDeleted = true;
+                } else {
+                    LOGGER.info("Cannot delete the Cloudformation because it contains instances");
+                }
+            } else {
+                LOGGER.info("Cloud formation does not exist on AWS side, maybe deleted previously or by the customer");
+                cloudFormationTemplateDeleted = true;
+            }
```

Comment:
instead of initializing the variable here with false, could you move this out to a separate method and return in every branch with the right value?

Reply:
what does it mean `branch`?

Reply:
if-else branches

Reply:
No, I prefer this way of return as well